src/coverlet.core/Instrumentation/ModuleTrackerTemplate.cs

```diff
@@ -73,11 +74,42 @@ public static void RecordSingleHit(int hitLocationIndex)
                 location = 1;
         }
 
-        public static void UnloadModule(object sender, EventArgs e)
+        private static bool IsDotNetCore()
+        {
+            // object for .NET Framework is inside mscorlib.dll
+            return Path.GetFileName(typeof(object).Assembly.Location) == "System.Private.CoreLib.dll";
+        }
+
+        public static void ProcessExitEvent(object sender, EventArgs e)
+        {
+            if (IsCalledByInProcessCollector && IsDotNetCore())
```

Comment:
At least for my test projects, moving the call to `IsDotNetCore` to https://github.com/tonerdo/coverlet/pull/754/files#diff-164040e68c2b496a18449c2363998152L111 still solves my issue - and should address your concern about people who don't update their test sdk (https://github.com/tonerdo/coverlet/issues/736#issuecomment-592154536).

Reply:
Basically, we could remove this whole if block, and replace 
`if (failedToCreateNewHitsFile)` with `if (failedToCreateNewHitsFile && !IsDotNetCore())`
It might be a safer change?

Reply:
> At least for my test projects, moving the call to IsDotNetCore to https://github.com/tonerdo/coverlet/pull/754/files#diff-164040e68c2b496a18449c2363998152L111 still solves my issue - and should address your concern about people who don't update their test sdk (#736 (comment)).

Cannot understand where this link pointing...can you rewrite it?

Reply:
> if (failedToCreateNewHitsFile) with if (failedToCreateNewHitsFile && !IsDotNetCore())

This could work...I'm only scared about remove redundant "flush" on exit(I want to do it for a long time) because could hidden some latent bug.
I'll take a look asap, busy weeks for me.
Thank's for investigation.

Reply:
> > At least for my test projects, moving the call to IsDotNetCore to https://github.com/tonerdo/coverlet/pull/754/files#diff-164040e68c2b496a18449c2363998152L111 still solves my issue - and should address your concern about people who don't update their test sdk (#736 (comment)).
> 
> Cannot understand where this link pointing...can you rewrite it?

Updated it to point at the file outside this pr

Reply:
Understood now.

Reply:
The other thing I noticed is that the process that updates the hits file is pretty inefficient (seeking back and forth in the file is slow from what I understand). I was able generate coverage successfully after rewriting it as follows (this could probably be further optimized by reading/writing more than one `int` at a time):
```cs
        private static async System.Threading.Tasks.Task UpdateHits(int[] hitsArray)
        {
            var readBuffer = new byte[sizeof(int)];
            var writeBuffer = new byte[sizeof(int)];

            using (var readStream = new FileStream(HitsFilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
            using (var writeStream = new FileStream(HitsFilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
            {
                int hitsLength = await ReadNext();
                var readTask = ReadNext();
                WriteLog($"Current hits found '{hitsLength}'");

                if (hitsLength != hitsArray.Length)
                {
                    throw new InvalidOperationException(
                        $"{HitsFilePath} has {hitsLength} entries but on memory {nameof(HitsArray)} has {hitsArray.Length}");
                }
                writeStream.Seek(sizeof(int), SeekOrigin.Current);
                for (int i = 0; i < hitsLength; ++i)
                {
                    int hits = await readTask;
                    readTask = ReadNext(); // start reading the next hits before we wait to write

                    if (SingleHit)
                        hits = hitsArray[i] + hits > 0 ? 1 : 0;
                    else
                        hits = hitsArray[i] + hits;
                    
                    await Write(hits);
                }

                async System.Threading.Tasks.Task<int> ReadNext()
                {
                    var bytesRead = await readStream.ReadAsync(readBuffer, 0, sizeof(int));
                    if (bytesRead < sizeof(int))
                    {
                        return -1;
                    }

                    return BitConverter.ToInt32(readBuffer, 0);
                }

                async System.Threading.Tasks.Task Write(int hits)
                {
                    writeBuffer[0] = (byte)hits;
                    writeBuffer[1] = (byte)(hits >> 8);
                    writeBuffer[2] = (byte)(hits >> 16);
                    writeBuffer[3] = (byte)(hits >> 24);
                    await writeStream.WriteAsync(writeBuffer, 0, sizeof(int));
                }
            }
        }
```

Reply:
For perf you should open a new issue and we'll tag with proper label, also every perf update must be supported by perf testing in test project https://github.com/tonerdo/coverlet/tree/master/test/coverlet.core.performancetest(we should use BDN)

Reply:
I was just trying to offer an alternative to not writing the hits file on process exit 😄 - if you'd like me to submit a separate PR with this change instead, I can do so.

Reply:
yep yep I'm ok for help and update...but we need to be very very careful coverlet is used by a lot of people and I try to fix bug and add useful feat before perf and to be honest write file is not a great concern at the moment.
This issue should have more impact but it's very challenging https://github.com/tonerdo/coverlet/issues/307

https://nugettrends.com/packages?months=12&ids=coverlet.msbuild&ids=coverlet.collector&ids=coverlet.console

Reply:
My only reason for looking at it is that when the file is not modified, or it is written quickly, then I'm able to successfully generate coverage. When it is slow, my workflow will often fail to generate coverage. I agree that your cautious approach is the correct one, which is also my motivation for thinking of alternatives to the initial approach of not attaching the process exit event handler.
And I do want to make sure you know how much I appreciate your help with this issue! I've tested all the solutions we've discussed with my real project that currently can't generate coverage, and am happy to contribute in a way that makes the most sense for the coverlet repo (even if that is just waiting for you to have a chance to test in the coming weeks).

Reply:
It's not clear to me why the "synch with mutex on get reports to protect hit file" doesn't work, that should be the simpler and less intrusive one.

Reply:
That is an interesting one - iirc the [repro repo](https://github.com/namely/CoverletErrorRepro) I gave you will exhibit the failure to generate coverage with that approach. Might make it easier to test why?

Reply:
will do asap

Reply:
I cannot repro on my local...no exception, do you have some advice or variation that could help to repro?

Reply:
Huh, that is odd. Just to make sure we're on the same page - this didn't produce an exception when running in `docker-compose`? Running in docker is essentially the only time I see the exception and/or coverage failing to generate.

Reply:
And to hopefully add some clarification, I added a readme to that repo.
Let me try to pull your change with the mutex and build it in that repo (in a different branch) so I can demonstrate the issue.

Reply:
Right, got it to fail intermittently using the mutex commit: https://github.com/namely/CoverletErrorRepro/tree/use-coverlet-mutex
As stated in the readme, this fails roughly 1/3 times on my laptop. I've noticed our CI environment is much slower than my laptop at running tests, so it is perhaps exacerbating the problem.

Reply:
Thank's for sample and for correct exception, that fix doesn't work my fault...synch is working but if thread acquired mutex inside tracker is killed wait on get coverage will fail for abandoned mutex and also if we catch that exception we cannot be sure that file was fully written so we can hit corrupted hit file.

The unique clean solution is the one in full patch or your idea of skip re-write if we're in .net core app because if we already called flush from collector the second one should be process exits(no app domain so no intermediate flushes expected and only one file creation expected for msbuild/.net tool with .net core runtime).

Reply:
No problem! Let me know if I can help with this in any way.

Reply:
Now we've a complete analysis, I'll think about best and light solution, for now you can fork and use your idea.